File: sdk/typescript/src/key-rotation.ts
import {
  U32_MAX,
  encodeChainId,
  encodeText,
  validateZeroneAddress,
  writeUint32,
} from "./auth-signing-common";

export const KEY_ROTATION_AUTHORIZATION_DOMAIN =
  "zerone.auth/rotate-key/v1" as const;
export const KEY_ROTATION_ACCEPTANCE_DOMAIN =
  "zerone.auth/accept-key/v1" as const;
export const KEY_ROTATION_AUTHORIZATION_MAX_TTL_SECONDS = 600n;

const I64_MAX = 0x7fff_ffff_ffff_ffffn;

export interface KeyRotationAuthorization {
  readonly chainId: string;
  readonly sender: string;
  readonly currentKeyVersion: number;
  readonly newOperationalKey: Uint8Array;
  readonly authorizationExpiresAtUnix: bigint;
}

/**
 * Returns the exact domain-separated bytes signed by the current Ed25519
 * operational key for MsgRotateKey. The chain verifies the expiry against
 * consensus block time; caller wall-clock time is not authoritative.
 */
export function keyRotationAuthorizationSignBytes(
  authorization: KeyRotationAuthorization,
): Uint8Array {
  return keyRotationSignBytes(
    KEY_ROTATION_AUTHORIZATION_DOMAIN,
    authorization,
  );
}

/**
 * Returns the exact domain-separated bytes signed by the proposed new
 * Ed25519 operational key. This proof is distinct from the current key's
 * authorization while binding the same transition fields.
 */
export function keyRotationAcceptanceSignBytes(
  authorization: KeyRotationAuthorization,
): Uint8Array {
  return keyRotationSignBytes(KEY_ROTATION_ACCEPTANCE_DOMAIN, authorization);
}

function keyRotationSignBytes(
  domainName: string,
  authorization: KeyRotationAuthorization,
): Uint8Array {
  const chainId = encodeChainId(authorization.chainId);
  validateZeroneAddress(authorization.sender);
  const sender = encodeText(authorization.sender, "sender");
  if (
    !Number.isInteger(authorization.currentKeyVersion) ||
    authorization.currentKeyVersion <= 0 ||
    authorization.currentKeyVersion > U32_MAX
  ) {
    throw new RangeError("current key version must be a positive uint32");
  }
  if (authorization.newOperationalKey.length !== 32) {
    throw new RangeError("new operational key must be 32 bytes");
  }
  if (
    authorization.authorizationExpiresAtUnix <= 0n ||
    authorization.authorizationExpiresAtUnix > I64_MAX
  ) {
    throw new RangeError(
      "authorization expiry must be a positive signed int64 Unix timestamp",
    );
  }

  const domain = new TextEncoder().encode(domainName);
  const result = new Uint8Array(
    domain.length + 1 + 4 + chainId.length + 4 + sender.length + 4 + 8 + 32,
  );
  let offset = 0;
  result.set(domain, offset);
  offset += domain.length;
  result[offset] = 0;
  offset += 1;
  offset = writeUint32(result, offset, chainId.length);
  result.set(chainId, offset);
  offset += chainId.length;
  offset = writeUint32(result, offset, sender.length);
  result.set(sender, offset);
  offset += sender.length;
  offset = writeUint32(result, offset, authorization.currentKeyVersion);
  new DataView(result.buffer, result.byteOffset, result.byteLength).setBigInt64(
    offset,
    authorization.authorizationExpiresAtUnix,
    false,
  );
  offset += 8;
  result.set(authorization.newOperationalKey, offset);
  return result;
}

File: x/auth/types/key_rotation.go
package types

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"strings"
	"time"
)

const (
	// KeyRotationAuthorizationDomain separates operational-key proofs from
	// every other Ed25519 signature accepted by Zerone.
	KeyRotationAuthorizationDomain = "zerone.auth/rotate-key/v1"
	// KeyRotationAcceptanceDomain separates proof by the proposed new key from
	// authorization by the current key, while binding the same transition.
	KeyRotationAcceptanceDomain = "zerone.auth/accept-key/v1"

	// KeyRotationAuthorizationMaxTTL bounds how far the signed expiry may be in
	// the future relative to consensus block time. Without a signed-at value or
	// unpredictable chain challenge, it does not prove when a signature was made.
	// The transaction remains independently subject to Cosmos sequence and
	// timeout-height rules.
	KeyRotationAuthorizationMaxTTL = 10 * time.Minute
)

// KeyRotationAuthorizationSignBytes returns the exact bytes that the current
// operational Ed25519 key must sign. The current key version is a single-use
// nonce, chainID prevents cross-chain replay, and expiresAtUnix is checked
// against consensus block time by the message server.
//
// Encoding (all integers big endian):
//
//	domain || 0x00 || u32(len(chain_id)) || chain_id ||
//	u32(len(sender)) || sender || u32(current_key_version) ||
//	i64(authorization_expires_at_unix) || new_key[32]
func KeyRotationAuthorizationSignBytes(
	chainID string,
	sender string,
	currentKeyVersion uint32,
	newOperationalKey []byte,
	expiresAtUnix int64,
) ([]byte, error) {
	return keyRotationSignBytes(
		KeyRotationAuthorizationDomain,
		chainID,
		sender,
		currentKeyVersion,
		newOperationalKey,
		expiresAtUnix,
	)
}

// KeyRotationAcceptanceSignBytes returns the exact bytes the proposed new
// operational key must sign to confirm possession and acceptance of the same
// transition authorized by the current key.
func KeyRotationAcceptanceSignBytes(
	chainID string,
	sender string,
	currentKeyVersion uint32,
	newOperationalKey []byte,
	expiresAtUnix int64,
) ([]byte, error) {
	return keyRotationSignBytes(
		KeyRotationAcceptanceDomain,
		chainID,
		sender,
		currentKeyVersion,
		newOperationalKey,
		expiresAtUnix,
	)
}

func keyRotationSignBytes(
	domainName string,
	chainID string,
	sender string,
	currentKeyVersion uint32,
	newOperationalKey []byte,
	expiresAtUnix int64,
) ([]byte, error) {
	if chainID == "" || chainID != strings.TrimSpace(chainID) {
		return nil, fmt.Errorf("chain ID must be non-empty without surrounding whitespace")
	}
	if err := ValidateCanonicalAccountAddress(sender); err != nil {
		return nil, fmt.Errorf("invalid sender address: %w", err)
	}
	if currentKeyVersion == 0 {
		return nil, fmt.Errorf("current key version must be positive")
	}
	if len(newOperationalKey) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("new operational key must be %d bytes", ed25519.PublicKeySize)
	}
	if expiresAtUnix <= 0 {
		return nil, fmt.Errorf("authorization expiry must be a positive Unix timestamp")
	}

	domain := []byte(domainName)
	chain := []byte(chainID)
	address := []byte(sender)
	if uint64(len(chain)) > uint64(^uint32(0)) || uint64(len(address)) > uint64(^uint32(0)) {
		return nil, fmt.Errorf("chain ID or sender is too long")
	}

	capacity := len(domain) + 1 + 4 + len(chain) + 4 + len(address) + 4 + 8 + ed25519.PublicKeySize
	result := make([]byte, 0, capacity)
	result = append(result, domain...)
	result = append(result, 0)
	result = appendUint32(result, uint32(len(chain)))
	result = append(result, chain...)
	result = appendUint32(result, uint32(len(address)))
	result = append(result, address...)
	result = appendUint32(result, currentKeyVersion)
	result = binary.BigEndian.AppendUint64(result, uint64(expiresAtUnix))
	result = append(result, newOperationalKey...)
	return result, nil
}

// OperationalKeyHash returns the canonical lowercase SHA-256 commitment for
// a 32-byte operational public key.
func OperationalKeyHash(key []byte) (string, error) {
	if len(key) != ed25519.PublicKeySize {
		return "", fmt.Errorf("operational key must be %d bytes", ed25519.PublicKeySize)
	}
	digest := sha256.Sum256(key)
	return hex.EncodeToString(digest[:]), nil
}

func appendUint32(dst []byte, value uint32) []byte {
	return binary.BigEndian.AppendUint32(dst, value)
}
